src/lib/profiles/data-management/Current/SubscriptionEngine.cpp

```diff
@@ -1515,7 +1520,675 @@ void SubscriptionEngine::OnCustomCommand(nl::Weave::ExchangeContext * aEC, const
+                                                      uint32_t aCurrentIndex)
+{
+    uint32_t profileId  = 0;
+    uint16_t statusCode = 0;
+
+    if (WEAVE_ERROR_ACCESS_DENIED == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_Common;
+        statusCode = Weave::Profiles::Common::kStatus_AccessDenied;
+        err        = WEAVE_NO_ERROR;
+    }
+    else if (WEAVE_ERROR_INVALID_PROFILE_ID == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_WDM;
+        statusCode = kStatus_InvalidPath;
+        err        = WEAVE_NO_ERROR;
+    }
+    else if (WEAVE_ERROR_WDM_VERSION_MISMATCH == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_WDM;
+        statusCode = kStatus_VersionMismatch;
+        err        = WEAVE_NO_ERROR;
+    }
+    else if (WEAVE_NO_ERROR == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_Common;
+        statusCode = Weave::Profiles::Common::kStatus_Success;
+    }
+    else
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_Common;
+        statusCode = Weave::Profiles::Common::kStatus_InternalError;
+    }
+
+    StatusDataHandleElement * statusDataHandleList =
+        reinterpret_cast<StatusDataHandleElement *> WEAVE_SYSTEM_ALIGN_SIZE((size_t)(pBuf->Start()), 4);
+    statusDataHandleList[aCurrentIndex].mProfileId       = profileId;
+    statusDataHandleList[aCurrentIndex].mStatusCode      = statusCode;
+    statusDataHandleList[aCurrentIndex].mTraitDataHandle = aTraitDataHandle;
+}
+
+/**
+ * Check if this triat path is the starting one
+ */
+bool SubscriptionEngine::IsStartingPath(StatusDataHandleElement * apStatusDataHandleList, TraitDataHandle aTraitDataHandle,
+                                        uint32_t aCurrentIndex)
+{
+    bool isStarting = true;
+    // Todo: Optimize to reduce loop lookup
+    for (uint32_t index = 0; index < aCurrentIndex; index++)
+    {
+        if ((nl::Weave::Profiles::kWeaveProfile_Common == apStatusDataHandleList[index].mProfileId) &&
+            (nl::Weave::Profiles::Common::kStatus_Success == apStatusDataHandleList[index].mStatusCode) &&
+            (aTraitDataHandle == apStatusDataHandleList[index].mTraitDataHandle))
+        {
+            isStarting = false;
+        }
+    }
+    return isStarting;
+}
+
+/**
+ * Update all trait versions according to temporary statusDataHandleList, each trait only bump version at starting path
+ */
+WEAVE_ERROR SubscriptionEngine::UpdateTraitVersions(PacketBuffer * apBuf, const TraitCatalogBase<TraitDataSource> * apCatalog,
+                                                    uint32_t aNumDataElements)
+{
+    WEAVE_ERROR err                            = WEAVE_NO_ERROR;
+    TraitDataSource * dataSource               = NULL;
+    TraitUpdatableDataSource * updatableSource = NULL;
+    StatusDataHandleElement * statusDataHandleList =
+        reinterpret_cast<StatusDataHandleElement *> WEAVE_SYSTEM_ALIGN_SIZE((size_t)(apBuf->Start()), 4);
```

Comment:
Given this line is used somewhat often, is this an opportunity for a convenience macro?

Reply:
Done, thanks